### PR TITLE
improve validation

### PR DESCRIPTION
**Pull request by ObiWahn**

### CODE REVIEW

**Comment by ObiWahn** [comment]
Still need to add some more tests and a jenkins run.

File: 3rdParty/json-schema-validation/include/validation/events_from_slice.hpp [hunk 1]
```diff
@@ -79,14 +81,51 @@ void from_value(Consumer& consumer,
...
+                            if (!enum_contains_value(special, arangodb::validation::SpecialProperties::Rev)) {
+                                continue;
+                            }
+                            break;
+                        }
+                        case VelocyPackHelper::FromAttribute:
+                            if (!enum_contains_value(special, arangodb::validation::SpecialProperties::From)) {
+                                continue;
+                            }
+                            break;
```
Comment (jsteemann) [inline 3rdParty/json-schema-validation/include/validation/events_from_slice.hpp]: out of curiosity, why isn't the `FromAttribute` case enclosed in a scope, but all other cases are?

File: 3rdParty/json-schema-validation/src/validation.cpp [hunk 2]
```diff
@@ -90,17 +90,55 @@ std::string const to_string = "_to";
...
 namespace arangodb::validation {
+SpecialProperties view_to_special(std::string_view view) {
+    if (view == key_string) {
+        return SpecialProperties::Key;
+    } else if (view == id_string) {
+        return SpecialProperties::Id;
+    } else if (view == rev_string) {
+        return SpecialProperties::Rev;
+    } else if (view == to_string) {
+        return SpecialProperties::From;
```
Comment (jsteemann) [inline 3rdParty/json-schema-validation/src/validation.cpp]:
```suggestion
        return SpecialProperties::To;
```

File: 3rdParty/json-schema-validation/src/validation.cpp [hunk 3]
```diff
@@ -90,17 +90,55 @@ std::string const to_string = "_to";
...
+    } else if (view == from_string) {
+        return SpecialProperties::To;
```
Comment (jsteemann) [inline 3rdParty/json-schema-validation/src/validation.cpp]:
```suggestion
        return SpecialProperties::From;
```

File: arangod/VocBase/Validators.cpp [hunk 4]
```diff
@@ -78,6 +96,24 @@ ValidatorBase::ValidatorBase(VPackSlice params)
...
+    if (specialSlice.compareString(StaticStrings::ValidatorPropertyNone) == 0) {
+      this->_special = validation::SpecialProperties::None;
+    } else if (specialSlice.compareString(StaticStrings::ValidatorPropertyAll) == 0) {
+      this->_special = validation::SpecialProperties::All;
+    } else {
+      using namespace std::literals::string_literals;
+      THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_VALIDATION_BAD_PARAMETER,
+                                     "Valid built-in attributes are: "s + StaticStrings::ValidatorPropertyNone +
+                                         ", " + StaticStrings::ValidatorPropertyAll +
+                                         "");
```
Comment (jsteemann) [inline arangod/VocBase/Validators.cpp]:
```suggestion
                                         ", " + StaticStrings::ValidatorPropertyAll);
```

File: arangod/VocBase/Validators.cpp [hunk 5]
```diff
@@ -48,10 +48,28 @@ std::string const&  to_string(ValidationLevel level) {
...
+  switch (special) {
+    case validation::SpecialProperties::All:
+      return StaticStrings::ValidatorPropertyAll;
+    case validation::SpecialProperties::None:
+      return StaticStrings::ValidatorPropertyNone;
+    case validation::SpecialProperties::Key:
+    case validation::SpecialProperties::Id:
+    case validation::SpecialProperties::Rev:
+    case validation::SpecialProperties::From:
+    case validation::SpecialProperties::To:
```
Comment (jsteemann) [inline arangod/VocBase/Validators.cpp]: is it intentional to make all these values produce a result of `ValidatorPropertyNone`?

Comment (ObiWahn) [reply arangod/VocBase/Validators.cpp]: Yes - For now it distinguished only between `none` and `all`. The rest defaults to `none`.

Comment (jsteemann) [reply arangod/VocBase/Validators.cpp]: Ok

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9063/

**Comment by ObiWahn** [comment]
We need a functionality that allows us to check all documents in a collection are valid. And in case there are invalid documents the keys should be returned.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9067/

File: arangosh/Import/ImportFeature.cpp [hunk 6]
```diff
@@ -180,6 +181,10 @@ void ImportFeature::collectOptions(std::shared_ptr<options::ProgramOptions> opti
   options->addOption(
       "--latency", "show 10 second latency statistics (values in microseconds)",
       new BooleanParameter(&_latencyStats));
+
+  options->addOption(
+      "--skip-validation", "skips document validation during import",
+      new BooleanParameter(&_skipValidation)).setIntroducedIn(370);
```
Comment (jsteemann) [inline arangosh/Import/ImportFeature.cpp]:
```suggestion
      new BooleanParameter(&_skipValidation)).setIntroducedIn(30700);
```

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9079/

**Pull request merged**